Sources/NIOSSL/ByteBufferBIO.swift

```diff
@@ -217,21 +217,33 @@ final class ByteBufferBIO {
             preconditionFailure("Unable to initialize custom BIO")
         }
 
-        // We now need to complete the BIO initialization. The BIO does not have an owned pointer
-        // to us, as that would create an annoying-to-break reference cycle.
+        // We now need to complete the BIO initialization. The BIO takes an owned reference to self here,
+        // which is broken on close().
         self.bioPtr = bio
-        CNIOBoringSSL_BIO_set_data(self.bioPtr, Unmanaged.passUnretained(self).toOpaque())
+        CNIOBoringSSL_BIO_set_data(self.bioPtr, Unmanaged.passRetained(self).toOpaque())
         CNIOBoringSSL_BIO_set_init(self.bioPtr, 1)
         CNIOBoringSSL_BIO_set_shutdown(self.bioPtr, 1)
     }
 
     deinit {
-        // On deinit we need to drop our reference to the BIO, and also ensure that it doesn't hold any
-        // pointers to this object anymore.
-        CNIOBoringSSL_BIO_set_data(self.bioPtr, nil)
+        // On deinit we need to drop our reference to the BIO.
         CNIOBoringSSL_BIO_free(self.bioPtr)
```

Comment:
could we assert we've been closed?

Reply:
prolly you can `CNIOBoringSSL_BIO_get_data(self.bioPtr) == nil` ?

Reply:
Yup, added an assertion.